pkg/volume/util/nested_volumes.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"k8s.io/api/core/v1"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// MakeNestedMountpoints creates mount points in baseDir for volumes mounted beneath name
+func MakeNestedMountpoints(name, baseDir string, pod v1.Pod) error {
+	checkContainer := func(container *v1.Container) error {
+		var mountPoints []string
+		var myMountPoints []string
+		for _, vol := range container.VolumeMounts {
+			cleaned := filepath.Clean(vol.MountPath)
+			mountPoints = append(mountPoints, cleaned)
+			if vol.Name == name {
+				myMountPoints = append(myMountPoints, cleaned)
+			}
+		}
+		sort.Strings(mountPoints)
+		parentPrefix := ".." + string(os.PathSeparator)
+		for _, mountPoint := range myMountPoints {
+			if strings.HasPrefix(mountPoint, parentPrefix) {
+				// Don't let a container trick us into creating directories outside of its rootfs
+				return fmt.Errorf("Invalid container mount point %v", mountPoint)
+			}
+			mpSlash := mountPoint + string(os.PathSeparator)
+			// cur represents the current mountpoint to check.
+			// The search makes it so cur starts at the first mount point after mountPoint
+			// The loop continues until it reaches the end of mountPoints or until cur references a mount point that is not nested beneath mountPoint
+			// prev represents the the last-encountered nested mountPoint. If set (i.e., not -1), it's used to compare with a mount point to make
+			//      sure the mount point is not a double-nested mount point
+			for prev, cur := -1, sort.SearchStrings(mountPoints, mpSlash); cur < len(mountPoints) && strings.HasPrefix(mountPoints[cur], mpSlash); cur++ {
```

Comment:
Would it be better to create a new filtered mount points list where all the entries have mpSlash as the prefix?

Could you potentially have a scenario where:
* mpSlash is in the mount point string, but is not the prefix
* that mount point comes before the real mpSlash mount points in sorted order, so this loop will break out early

Reply:
I'm not certain I understand your scenario, but if I do, I don't think it is possible only because mpSlash necessarily has a trailing slash.  So, for example, given `/mnt`, `/mnt/nested` and `/mnt_not_nested`, when `/mnt` is being prepared, the `SearchStrings()` call would return the first thing after `/mnt/`, which would be `/mnt/nested`. This is true whether `_` or `/` comes first lexicographically because of the appended `/`.  With that in mind, do you still think it's worth filtering the list before looping? (I'm fine either way).

Reply:
Ah sorry I thought SearchStrings would return the first string where it matched by subset.  But it does exact match, so this is fine.